[WIP] Use cache around magit-run-git calls

npostavs opened the pull request:
Continued from #3094.

This initial revision is just the simplest thing possible: extend the refresh cache around `magit-run-git`. We might get in trouble if the git command we run changes something we cache, I'm not sure if that can happen though? The main savings seems to be the `rev-parse` calls, so we could filter the cache to keep just those if it's a problem.

Here's the diff of git calls for a single `magit-stage`.
```diff
--- count-master.txt	2017-06-01 00:10:11.951496805 -0400
+++ count-cache2.txt	2017-06-01 00:11:07.457862768 -0400
@@ -22,27 +22,13 @@
 git show --no-patch --format=%s master^{commit} --
 git rev-parse --verify master
 git config --list -z
-git rev-parse --show-cdup
-git rev-parse --show-toplevel
 git show --no-patch --format=%h %s current^{commit} --
 git symbolic-ref --short HEAD
 git rev-parse --verify HEAD
 git update-index --refresh
-git rev-parse --show-toplevel
-git rev-parse --show-cdup
-git rev-parse --show-cdup
-git rev-parse --show-toplevel
 git add -u -- README.md
-git config -z --get-all magit.extension
-git rev-parse --show-toplevel
 git rev-parse --show-cdup
-git rev-parse --show-cdup
-git rev-parse --show-toplevel
-git rev-parse --show-toplevel
-git rev-parse --show-cdup
-git rev-parse --show-cdup
-git rev-parse --show-toplevel
 git rev-parse --show-toplevel
 
 
-45 calls in 0.168s
+31 calls in 0.147s
```

Code I used to measure:
```elisp
(defvar git-call-list t)
(defun around-magit-process-file (fun program &optional infile buffer display &rest args)
  (when (listp git-call-list)
    (push (cons program (mapcar #'substring-no-properties
                                (cl-set-difference args magit-git-global-arguments :test #'equal)))
          git-call-list))
  (let ((tbeg (current-time))
        (tend nil))
    (prog1 (apply fun program infile buffer display args)
      (setq tend (current-time))
      (when (listp git-call-list)
        (push (float-time (time-subtract tend tbeg))
              (car git-call-list))))))
(advice-add 'magit-process-file :around #'around-magit-process-file)

(defun toggle-magit-git-counting (enable)
  (interactive (list (not (listp git-call-list))))
  (if enable
      (setq git-call-list ())
    (with-current-buffer (get-buffer-create "*magit calls*")
      (if git-call-list
          (let ((total-time 0.0))
            (pcase-dolist (`(,time . ,call) git-call-list)
              (insert (format "|%.3fs" time) "|" (mapconcat #'identity call " ") "\n")
              (cl-incf total-time time))
            (insert (format "\n\n%d calls in %.3fs\n"
                            (length git-call-list) total-time)))
        (erase-buffer))
      (pop-to-buffer (current-buffer))
      (setq git-call-list t)))
  (message "%sabled git counting" (if enable "en" "dis")))

(setq magit-refresh-verbose t)
(global-set-key [f12] 'toggle-magit-git-counting)
```

tarsius commented:
Sorry for not replying earlier.

> We might get in trouble if the git command we run changes something we cache, I'm not sure if that can happen though?

Commands that create repositories could be a problem.

npostavs commented:
> Sorry for not replying earlier.

There's no rush.

> Commands that create repositories could cause problems.

Hmm, I added some debug code to the cache macro, but I couldn't get any errors using `magit-init`, `magit-clone`, or `magit-submodule-add` (via staging of a repo in a subdir).
```diff
modified   lisp/magit-git.el
@@ -221,6 +221,21 @@ (defcustom magit-list-refs-sortby nil
 ;;; Git
 
 (defvar magit--refresh-cache nil)
+(defvar magit--refresh-cache-debug t)
+
+(defconst hash-table-equal--sentinel (make-symbol "hash-table-equal--sentinel"))
+
+(defun hash-table-equal-p (h1 h2)
+  "Return non-nil if hash-table H1 has same keys & values as H2."
+  (and (= (hash-table-count h1)
+          (hash-table-count h2))
+       (catch 'equal
+         (maphash
+          (lambda (k v)
+            (unless (equal v (gethash k h2 hash-table-equal--sentinel))
+              (throw 'equal nil)))
+          h2)
+         t)))
 
 (defmacro magit--with-refresh-cache (key &rest body)
   (declare (indent 1) (debug (form body)))
@@ -229,6 +244,11 @@ (defmacro magit--with-refresh-cache (key &rest body)
          (let ((,k ,key))
            (--if-let (assoc ,k (cdr magit--refresh-cache))
                (progn (cl-incf (caar magit--refresh-cache))
+                      (when magit--refresh-cache-debug
+                        (let ((value ,(macroexp-progn body)))
+                          (cl-assert (if (hash-table-p value)
+                                         (hash-table-equal-p value (cdr it))
+                                       (equal value (cdr it))))))
                       (cdr it))
              (cl-incf (cdar magit--refresh-cache))
              (let ((value ,(macroexp-progn body)))
```

tarsius commented:
It was just a guess, I didn't look at any code before making it ;-)

tarsius commented:
Anyway, I am back home and will try to get back into the groove soon.

CyberShadow commented:
> Here's the diff of git calls for a single magit-stage.

Awesome!

I was wondering if it was possible to improve upon the ~30-second time to open `magit-status` in a sshfs-mounted directory, so I recorded a git execution trace as well:

```
[1499527946.198401083] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --show-toplevel
[1499527946.492438416] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --show-cdup
[1499527946.786543004] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --show-cdup
[1499527947.078818112] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --show-toplevel
[1499527947.370325349] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --show-toplevel
[1499527947.662528175] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --show-cdup
[1499527947.955700423] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --show-cdup
[1499527948.251836392] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --show-toplevel
[1499527948.823756245] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --is-bare-repository
[1499527949.117152045] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false config -z --get-all magit.extension
[1499527949.548829693] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --show-toplevel
[1499527949.887884677] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false update-index --refresh
[1499527950.549541483] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --verify HEAD
[1499527951.487731326] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false symbolic-ref --short HEAD
[1499527952.195825312] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false show --no-patch --format=%h %s master^{commit} --
[1499527953.733595930] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --show-toplevel
[1499527954.024175857] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --show-cdup
[1499527954.315711588] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false config --list -z
[1499527954.607370095] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --verify gitlab/master
[1499527955.451881670] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false show --no-patch --format=%s gitlab/master^{commit} --
[1499527956.898567528] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false describe --long --tags
[1499527957.834100379] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false describe --contains HEAD
[1499527959.323056596] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --git-dir
[1499527960.031331183] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false status -z --porcelain
[1499527962.627543242] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false diff --no-ext-diff --no-prefix --
[1499527964.070875263] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false diff --cached --no-ext-diff --no-prefix --
[1499527965.423918330] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --verify refs/stash
[1499527966.315534387] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse @{upstream}
[1499527967.944020527] git --no-pager -c core.preloadindex=true -c log.showSignature=false log --format=%h%d%x00%x00%aN%x00%at%x00%s --decorate=full -n256 --use-mailmap --no-prefix ..@{upstream} --
[1499527971.002124714] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse @{upstream}
[1499527972.674018735] git --no-pager -c core.preloadindex=true -c log.showSignature=false log --format=%h%d%x00%x00%aN%x00%at%x00%s --decorate=full -n256 --use-mailmap --no-prefix @{upstream}.. --
```

It looks like this patch should greatly improve the situation here as well!

kyleam commented:
FWIW, I merged this into my local integration branch a week or so ago
and, though I haven't tried specifically to break things, I haven't
noticed any issues.

tarsius commented:
And I have just studied the implementation once more and have concluded that it is safe to merge. @npostavs Could you please add a release note and then do so?

npostavs commented:
Done.

> I was wondering if it was possible to improve upon the ~30-second time to open magit-status in a sshfs-mounted directory,

@CyberShadow 30 seconds? Ouch. This patch will help a bit, but not that much. From your trace, I guess it should shave off about 4 seconds for you.

CyberShadow commented:
With Magit 20170711.1307:

```
[1499891665.889824802] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --show-toplevel
[1499891666.514659152] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --show-cdup
[1499891666.919327887] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --show-cdup
[1499891667.325289605] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --show-toplevel
[1499891667.730566701] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --show-toplevel
[1499891668.136678222] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --show-cdup
[1499891668.543857358] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --show-cdup
[1499891668.950963933] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --show-toplevel
[1499891669.616924978] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --is-bare-repository
[1499891670.021834794] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false config -z --get-all magit.extension
[1499891670.428998496] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --show-toplevel
[1499891670.835957852] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false update-index --refresh
[1499891672.109640877] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --verify HEAD
[1499891673.386673478] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false symbolic-ref --short HEAD
[1499891674.279378468] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false show --no-patch --format=%h %s master^{commit} --
[1499891676.612162738] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --show-toplevel
[1499891677.018410162] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --show-cdup
[1499891677.424633073] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false config --list -z
[1499891677.830024701] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --verify gitlab/master
[1499891678.799350079] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false show --no-patch --format=%s gitlab/master^{commit} --
[1499891680.380039981] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false describe --long --tags
[1499891681.437951394] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false describe --contains HEAD
[1499891683.110341832] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --git-dir
[1499891683.909443750] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false status -z --porcelain
[1499891687.321937365] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false diff --no-ext-diff --no-prefix --
[1499891689.298740402] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false diff --cached --no-ext-diff --no-prefix --
[1499891691.009118516] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --verify refs/stash
[1499891691.978843292] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse @{upstream}
[1499891693.819190790] git --no-pager -c core.preloadindex=true -c log.showSignature=false log --format=%h%d%x00%x00%aN%x00%at%x00%s --decorate=full -n256 --use-mailmap --no-prefix ..@{upstream} --
[1499891698.402159353] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse @{upstream}
[1499891700.245368463] git --no-pager -c core.preloadindex=true -c log.showSignature=false log --format=%h%d%x00%x00%aN%x00%at%x00%s --decorate=full -n256 --use-mailmap --no-prefix @{upstream}.. --
```

Doesn't seem like anything changed at all? I can see in ~.emacs.d/elpa/magit-20170711.1307/magit-mode.el that the patch is applied. Have I missed something?

npostavs commented:
Hmm, it might not be triggering for the log command properly, can you confirm that staging and unstaging at least don't call `rev-parse --show-cdup` and `rev-parse --show-toplevel` multiple times?

CyberShadow commented:
> can you confirm that staging and unstaging at least don't call rev-parse --show-cdup and rev-parse --show-toplevel multiple times?

Yes, seems like it doesn't:

```
[1499893815.902393695] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --show-toplevel
[1499893816.353126444] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --show-cdup
[1499893817.020197603] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false config --list -z
[1499893817.427653954] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false add -u -- notes.org
[1499893819.007907338] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false update-index --refresh
[1499893820.370217852] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --verify HEAD
[1499893821.476645003] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false symbolic-ref --short HEAD
[1499893822.448073060] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false show --no-patch --format=%h %s master^{commit} --
[1499893824.727235506] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --verify gitlab/master
[1499893825.610732513] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false show --no-patch --format=%s gitlab/master^{commit} --
[1499893827.110197181] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false describe --long --tags
[1499893827.907774124] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false describe --contains HEAD
[1499893829.844032356] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --git-dir
[1499893830.641300265] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false status -z --porcelain
[1499893834.495166479] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false diff --no-ext-diff --no-prefix --
[1499893835.207140841] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false diff --cached --no-ext-diff --no-prefix --
[1499893837.792139000] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse --verify refs/stash
[1499893838.813158556] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse @{upstream}
[1499893840.653250573] git --no-pager -c core.preloadindex=true -c log.showSignature=false log --format=%h%d%x00%x00%aN%x00%at%x00%s --decorate=full -n256 --use-mailmap --no-prefix ..@{upstream} --
[1499893844.826923456] git --no-pager --literal-pathspecs -c core.preloadindex=true -c log.showSignature=false rev-parse @{upstream}
[1499893847.615023197] git --no-pager -c core.preloadindex=true -c log.showSignature=false log --format=%h%d%x00%x00%aN%x00%at%x00%s --decorate=full -n256 --use-mailmap --no-prefix @{upstream}.. --
```

FWIW, the logs in my above comments were recorded by binding `magit-status` to `M-g` and pressing it.

tarsius commented:
In your latest log I only see those called once each. As for the other log, do you have the wip modes? Those are not covered by the cache yet.

CyberShadow commented:
I had not heard about wip modes before your comment, so no :)

npostavs commented:
I'm seeing some cache entries not being used when calling `magit-refresh` in the status buffer due to `default-directory` sometimes being abbreviated (e.g. `~/src/foo`) and sometimes not (`/home/user/src/foo`). The following fixes it, but I only got 2 duplicates, not 5 like in @CyberShadow's trace, so we might need some more investigation.

```diff
--- i/lisp/magit-git.el
+++ w/lisp/magit-git.el
@@ -286,7 +286,7 @@ (defun magit-git-str (&rest args)
 newline, return an empty string.  Like `magit-git-string' but
 ignore `magit-git-debug'."
   (setq args (-flatten args))
-  (magit--with-refresh-cache (cons default-directory args)
+  (magit--with-refresh-cache (cons (expand-file-name default-directory) args)
     (with-temp-buffer
       (apply #'magit-process-file magit-git-executable nil (list t nil) nil
              (magit-process-git-arguments args))
@@ -347,7 +347,7 @@ (defun magit-git-string (&rest args)
 If there is no output, return nil.  If the output begins with a
 newline, return an empty string."
   (setq args (-flatten args))
-  (magit--with-refresh-cache (cons default-directory args)
+  (magit--with-refresh-cache (cons (expand-file-name default-directory) args)
     (with-temp-buffer
       (apply #'magit-git-insert args)
       (unless (bobp)
@@ -428,7 +428,7 @@ (defun magit-git-dir (&optional path)
 All symlinks are followed.  If optional PATH is non-nil, then
 it has to be a path relative to the control directory and its
 absolute path is returned."
-  (magit--with-refresh-cache (list default-directory 'magit-git-dir path)
+  (magit--with-refresh-cache (list (expand-file-name default-directory) 'magit-git-dir path)
     (magit--with-safe-default-directory nil
       (--when-let (magit-rev-parse-safe "--git-dir")
         (setq it (file-name-as-directory (magit-expand-git-file-name it)))
@@ -469,7 +469,7 @@ (defun magit-toplevel (&optional directory)
 located within the working tree, then it is not possible to avoid
 returning the truename."
   (magit--with-refresh-cache
-      (cons (or directory default-directory) 'magit-toplevel)
+      (cons (expand-file-name (or directory default-directory)) 'magit-toplevel)
     (magit--with-safe-default-directory directory
       (-if-let (topdir (magit-rev-parse-safe "--show-toplevel"))
           (let (updir)
```

npostavs commented:
> I'm seeing some cache entries not being used when calling magit-refresh in the status buffer due to default-directory sometimes being abbreviated (e.g. ~/src/foo) and sometimes not (/home/user/src/foo).

Hmm, I'm not seeing this on my GNU/Linux box, could this be a side effect of filename conversions on Windows?

Anyway, the actual problem seems to be that `magit-status` in particular doesn't use the cache, see #3128.
